GoRogue/MapGeneration/Region.cs

```diff
@@ -0,0 +1,374 @@
+            InnerPoints = InnerFromOuterPoints(OuterPoints).Distinct().ToList();
+        }
+        #endregion
+
+        #region utilities
+        /// <summary>
+        /// Gets all Points that overlap with another region
+        /// </summary>
+        /// <param name="other">The region to evaluate against</param>
+        /// <returns>All overlapping points</returns>
+        public IEnumerable<Point> Overlap(Region other)
+        {
+            foreach (Point c in InnerPoints)
+            {
+                if (other.InnerPoints.Contains(c))
+                    yield return c;
+            }
+        }
+
+        /// <summary>
+        /// Whether or not this region contains a point
+        /// </summary>
+        /// <param name="here">The Point to evaluate</param>
+        /// <returns></returns>
+        public bool Contains(Point here)
+        {
+            return InnerPoints.Contains(here);
+        }
+        /// <summary>
+        /// Shifts the entire region by performing Point addition
+        /// </summary>
+        /// <param name="distance">the distance and direction to shift the region</param>
+        /// <returns></returns>
+        public Region Shift(Point distance)
+        {
+            Region region = new Region(Name, SouthEastCorner + distance, NorthEastCorner + distance, NorthWestCorner + distance, SouthWestCorner + distance);
+            foreach (var subRegion in SubRegions)
+            {
+                ((List<Region>) region.SubRegions).Add(subRegion);
+            }
+            return region;
+        }
+
+
+        /// <summary>
+        /// Cuts Points out of a parent region if that point exists in a subregion,
+        /// and cuts Points out of subregions if that point is already in a subregion
+        /// </summary>
+        /// <remarks>
+        /// This favors regions that were added later, like a stack.
+        /// </remarks>
+        public void DistinguishSubRegions()
+        {
+            List<Region> regionsDistinguished = new List<Region>();
+            IEnumerable<Region> regionsReversed = SubRegions;
+            regionsReversed.Reverse();
+            foreach (Region region in regionsReversed)
+            {
+                foreach (Point point in region.Points.Distinct())
+                {
+                    foreach (Region distinguishedRegion in regionsDistinguished)
+                    {
+                        if (distinguishedRegion.Contains(point))
+                        {
+                            while (region.OuterPoints.Contains(point))
+                                ((List<Point>) region.OuterPoints).Remove(point);
+
+                            while (region.InnerPoints.Contains(point))
+                                ((List<Point>) region.InnerPoints).Remove(point);
+                        }
+                    }
+                }
+
+                regionsDistinguished.Add(region);
+            }
+        }
+
+        /// <summary>
```

Comment:
This is probably a future-work item if anything; but this might actually be able to be pulled out of `Region` entirely and into a `GenerationStep` in map gen that deals with regions.  This may require some refactoring of how things work over there though.  I think this makes a great use-case to have in mind while designing the map generation algorithms in any case; but it can be left here for the time being.

Reply:
Yeah, its outside of scope for this effort. maybe next iteration